TopicWriter retry

sultanowskii opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull request type

<!-- Please try to limit your pull request to one types, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

- Moved

Issue Number: N/A

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

sultanowskii committed:
implement retry for TopicWriter, extract Retrier into a separate file

sultanowskii committed:
revert local testing leftover

sultanowskii committed:
fix format

sultanowskii committed:
fix linter issues

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -264,21 +274,57 @@ impl TopicWriter {
-                    })),
-                })
-                .unwrap(); // TODO: HANDLE ERROR
+            return match TopicWriter::retry(retrier, || async {
```
The retry won't work this way.
If sending a message results in an error, the topic writer must reconnect to the server before retrying.

rekby reviewed (changes requested):
Thanks for the PR. The retry must be implemented in a different way; I added comments in the file.

sultanowskii committed:
implement reconnect

sultanowskii committed:
fix linter issues

sultanowskii committed:
also update stream for receiver loop

sultanowskii committed:
wip: extract loops into a separate rebuildable structure

sultanowskii committed:
move write&exit request logic to Reconnector

sultanowskii committed:
todo

sultanowskii committed:
state

sultanowskii committed:
resolve more issues

sultanowskii committed:
fix compilation errors & several bugs

sultanowskii committed:
retrieve from init_response, fix other issues

sultanowskii committed:
(wip) fix fmt/linter issues, reorder args

sultanowskii committed:
remove unused functions, move logic

sultanowskii committed:
extract reconnector loop spawn into a separate function

sultanowskii committed:
remove mutex on streams entirely

sultanowskii committed:
remove excessive newline

sultanowskii committed:
fmt fix

sultanowskii committed:
unify a bit, add more TODOs

sultanowskii committed:
fix lock-related issues

sultanowskii committed:
remove excessive clone()s, use tokio mutex on messages

sultanowskii committed:
get rid of preventive messages clone()

sultanowskii committed:
fix style, names, typos

sultanowskii committed:
fix typo

sultanowskii committed:
fix logic

sultanowskii committed:
fix bug in stop

sultanowskii committed:
simplify comment

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+
+        let writer_state = Arc::new(Mutex::new(TopicWriterState::Working));
+
+        let (initial_messages_sender, initial_messages_receiver) = mpsc::channel(32_usize);
```
Почему тут размер канала именно 32?

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+                    want_reconnect_sender,
+                )
+                .await
+                {
```
Не нужно делать такие сложные вызовы между match и вариантами - такой код сложно читать.

Лусше лишний раз присвоить переменной промежуточное значение.

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -263,20 +612,52 @@ impl TopicWriter {
+            // - If write fails, then we put these unwritten messages back WITH taking into account messages that might've been written into messages.
+            //
+            // The "clone, clear if success" approach is dangerous because we can lose messages that are appended in another coroutine.
+            messages_guard.drain(..).collect::<Vec<MessageData>>()
```
**[Критично] Сообщения, уже отправленные в gRPC-поток, теряются при реконнекте**

Здесь сообщения изымаются из буфера через `drain()`. Если `send()` в gRPC-поток прошёл успешно (вернул `Ok`), они удаляются из `messages`. Но если соединение обрывается **после** отправки, но **до** получения ACK — эти сообщения теряются навсегда: их нет ни в `messages`, ни где-либо ещё.

При реконнекте сервер возвращает `init_response.last_seq_no` — всё с seq_no > last_seq_no нужно переслать. Но данных нет. В итоге клиент будет ждать ACK, который никогда не придёт, и `flush()` зависнет.

**Решение:** хранить sent-but-not-acked сообщения отдельно. При реконнекте восстанавливать сообщения с seq_no > server's last_seq_no для повторной отправки.

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+                            }
+                        };
+
+                        if TopicWriter::is_retry_allowed(&err) {
```
**[Критично] `writer_state` не сбрасывается в `Working` перед реконнектом**

Когда `writer_loop` получает ошибку, он устанавливает `*writer_state = FinishedWithError(...)` и отправляет ошибку в `want_reconnect_sender`. Здесь `reconnection_loop` принимает решение о ретрае и создаёт новый `WriteSupervisor` — но `writer_state` **по-прежнему** остаётся `FinishedWithError`.

В результате любой вызов `write_message()` во время реконнекта или после его успешного завершения вернёт ошибку через `is_cancelled()`, хотя подключение уже восстановлено.

Нужно сбрасывать `writer_state` в `Working` перед началом новой итерации подключения.

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+                                    message_receive_loop_writer_state.lock().unwrap(); // TODO handle error
                                 *writer_state =
-                                    TopicWriterMode::FinishedWithError(receive_message_iteration_error);
+                                    TopicWriterState::FinishedWithError(receive_message_iteration_error);
```
**[Критично] Ошибка в `receive_messages_loop` не вызывает реконнект**

Когда `receive_messages_loop_iteration` возвращает ошибку:
1. Отменяется `cancellation_token` супервайзера.
2. `writer_state` устанавливается в `FinishedWithError`.
3. `writer_loop` замечает отмену и завершается **без отправки** в `want_reconnect_tx`.

В итоге `want_reconnect_receiver` в `reconnection_loop` получает `Err(RecvError)` (канал закрыт), что обрабатывается как фатальное завершение — реконнект **не происходит**.

Обрыв соединения на приёмной стороне — самый частый сценарий — должен тоже вызывать переподключение. Нужно либо отправлять ошибку через `want_reconnect_tx` из receive_loop, либо обрабатывать `Err(RecvError)` в reconnection_loop как сигнал к реконнекту.

sultanowskii replied:
Ой, правда. Добавил прокидывание ошибок и в этом цикле, в обоих местах привел к одному виду

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+                                trace!("Channel error: {}", chan_err);
+                                let _ = supervisor.stop().await;  // TODO: handle error
+                                let mut writer_state = params.writer_state.lock().unwrap();
+                                *writer_state = TopicWriterState::FinishedWithError(
```
Тут получается проблема: если во время реконнекта - кто-то попробует отправить сообщение - он получит ошибку, т.к. статус писателя - завершён с ошибкой. В целевом состоянии ошибки на запись  нужно возвращать только если ошибка неретраибельная.

sultanowskii replied:
Добавил отдельное состояние

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+                )
+                .await
+                {
+                    Ok(supervisor) => supervisor,
```
Тут после успешного переподключение состояние писателя не меняется на "работает", а остаётся ошибкой. т.е. даже после реконнекта клиент ничего не сможет отправить.

rekby reviewed (changes requested):
PR движется в правильном направлении: архитектура с внешним `reconnection_loop` + внутренним `WriteSupervisor` логична. Но есть несколько критических багов, из-за которых ретраи либо не происходят вовсе, либо приводят к потере данных.

Кроме того, PR не содержит тестов на поведение при разрыве соединения. Хотелось бы видеть как минимум:
- тест на корректный реконнект при ретраибельной ошибке;
- тест на то, что сообщения не теряются при переподключении;
- тест на корректную остановку через `stop()`.

Плюс написал некоторые пожелания по неймингу

sultanowskii committed:
rename last_seq_num_handled to last_seq_no_assigned

sultanowskii committed:
rename init_state to connection_info

sultanowskii committed:
rename want_receonnect* to error*

sultanowskii committed:
set state to Working after successful reconnect

sultanowskii committed:
retry on idempotent-only errors

sultanowskii committed:
rename is_cancelled to check_working

sultanowskii committed:
fix typos in comment

sultanowskii committed:
rename WriteSupervisor to StreamWriter, PeriodicTaskParams to WriterLoopParams

sultanowskii committed:
enhance RetryDecision comment

sultanowskii committed:
fix stop() resource leak

sultanowskii committed:
make both loops send errors to error_tx, simplify flow

sultanowskii committed:
fix & use retry, fix typos and minor mistakes

sultanowskii commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+        trace!("Stopping...");
+
+        match timeout(self.flush_timeout, self.flush()).await {
+            Ok(result) => result?,
```
Сделал, теперь ошибки, если были, кидаются уже после всех завершающих операций

sultanowskii commented on ydb/src/retry.rs:
```diff
@@ -0,0 +1,64 @@
+use num::pow;
```
Занес Retry в writer. На текущий момент неясны два момента:
- сбрасывать ли attempt, если да, то в какой момент? Сразу после успешного создания StreamWriter?
- что использовать в качестве времени начала? Время первой неудачной попытки?

sultanowskii commented on ydb/src/retry.rs:
```diff
@@ -0,0 +1,64 @@
+        if params.time_from_start < self.timeout {
+            if params.attempt > 0 {
+                res.wait_timeout =
+                    Duration::from_millis(pow(INITIAL_RETRY_BACKOFF_MILLISECONDS, params.attempt));
```
Хорошее замечание, поправил, добавил верхний порог

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -85,105 +84,461 @@ impl Future for AckFuture {
+        sender
+            .send(message)
+            .await
+            .map_err(|err| YdbError::custom(format!("can't send the message to channel: {err}")))?;
```
В go/python я сделал так, что есть три уровня читателя и писателя: 
- верхнеуровневый интерфейс-декоратор. Сам ничего особо не делает, только переводит API удобное для пользователя в API, удобное внутри.
- реконнектор: хранит состояние между реконнекторами, управляет обработкой ошибок, реализует логику при реконнектах
- работа на стриме: нижний увовень. Задача перевести API, удобное внутри в вызовы GRPC-протокола. По любой ошибке просто падает и теряет своё состояние.

Тогда работа с сообщениями выглядит так:
1. клиент кладёт его в API вернхнего уровня
2. сообщение попадает в буфер реконнектора
3. из буфера реконнектора сообщения проталкиваются на уровень стрима.
4. уровень стрима отправляет обратно подтверждения
5. при ошибке уровень стрима закрывается и пересоздаётся. В него заново отправляются все накопленные в буфере реконнектора сообщения.

в такой системе не нужно что-то специально читить. Состояние, зависящее от стрима пропадает вместе с ним по любой ошибке. У меня так сделаны и читатель и писатель.

sultanowskii replied:
Похоже, поправил - теперь тикет добавляется в reception_queue на уровне StreamWriter, сразу после прихода сообщения от пользователя сверху.

Из-за этого в примере пришлось поменять `write` на `write_with_ack` - иначе, из-за того, что мы не дожидаемся последнего, не успевали отправить сообщение на сервер (т.к. батчевание и по таймауту не добежали)

sultanowskii committed:
don't update writer state in stream writer

sultanowskii committed:
extract stream writer into a separate file

sultanowskii committed:
extract reconnector

sultanowskii committed:
queue wip

sultanowskii committed:
using queue wip

sultanowskii committed:
enhance queue, use queue in writer

sultanowskii committed:
check seq_no in queue

sultanowskii committed:
actualize comments

sultanowskii committed:
remove completed todo

sultanowskii committed:
move adding ticket closer sending the message

sultanowskii committed:
write_with_ack -> write in example

sultanowskii committed:
add TODOs

sultanowskii committed:
add TODO

sultanowskii committed:
fix linter issues

rekby reviewed (commented):
И похоже у тебя получается оч. сложное устройство и синхронизации. Предлагаю остановиться, нарисовать как это всё должно работать, потом созвониться и обсудить. Я думаю что можно сделать проще, чтобы не путаться ни сейчас, ни потом при сопровождении кода.

sultanowskii committed:
move queue creation to the TopicWriter, add close method

sultanowskii committed:
rename method

sultanowskii committed:
improve message queue, add tests

sultanowskii committed:
group asserts better

sultanowskii committed:
make retry indefinite by default, calculate time_from_start in reconnector

sultanowskii committed:
simplify a bit

sultanowskii committed:
rearrange retry-related ops

sultanowskii commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,124 @@
+    ) -> Vec<MessageData> {
+        let length: usize = last_written_order_no as usize - *last_sent_order_no as usize;
+        let mut messages = Vec::with_capacity(length);
+        while *last_sent_order_no != last_written_order_no {
```
Map взял, подумав на листочке и посмотрев в гошную реализацию

Передаелал через VecDeque, выглядит неплохо

sultanowskii committed:
add reconnecting state

sultanowskii committed:
events-based wait/get_messages, various fixes

sultanowskii committed:
rename methods, simplify interface

sultanowskii committed:
get rid of process_new_messages_loop, simplify the whole structure

sultanowskii committed:
add indefinite reader for tests message_achkonwledged channel

sultanowskii committed:
add more queue tests

sultanowskii committed:
remove completed todo

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -381,65 +154,81 @@ impl TopicWriter {
             reception_queue.init_flush_op()?
         };
 
+        self.message_queue.close_for_new_messages();
```
Это получается что flush можно только один раз сделать?

flush() предназначен не для единичного сброса сообщений перед очисткой, а для того чтобы убедиться что всё что было записано до вызова - успешно сохранилось на сервер. При этом писатель остаётся работоспособным и в него можно продолжать писать сообщения, в т.ч. параллельно с тем как мы ждём результата flush (когда в будущем добавится параллельная запись).

sultanowskii replied:
Ух, да, flush как при работе с файлами, понял

Поправил, убрал отсюда закрытие очереди

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,602 @@
+use std::{
+    collections::VecDeque,
+    sync::{Arc, Mutex},
```
Внутри async-кода опасно использовать обычный mutex - код разных нитей может исполняться в одном реальном потоке. Тогда как только встретиться залоченный mutex остановится весь поток событий всех корутин и никто mutex освободить уже не сможет, даже если при реальном многопоточном коде всё ок.

Для синхронизации асинхронного кода посмотри на `tokio::sync::Mutex`.

sultanowskii replied:
Справедливо, поправил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,602 @@
+        inner.is_empty()
+    }
+
+    pub(crate) async fn wait(&self) {
```
Тут по названию непонятно какого события ждём

sultanowskii replied:
Уточнил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,602 @@
+    messages: VecDeque<MessageData>,
+
+    // index of the last message that has been added to the queue
+    last_written_index: Option<usize>,
```
Зачем нужен то индекс, то `seqno`? оба уникальные монотонно возрастающие. Будет ли проще если оставить что-то одно?

sultanowskii replied:
Придумал, как без него, вышло действительно проще, спасибо

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,602 @@
+            )));
+        }
+
+        self.messages.pop_front();
```
Тут можно сразу в начале делать pop_front и работать с извлечённым сообщением.

Если у нас что-то пошло не так - клиента надо закрыть. После поломки протокола он восстановиться уже не сможет, т.к. это внутренняя ошибка и непонятно что ещё пошло не так. Лучше и проще в таких случаях сразу останавливаться с ошибкой.

sultanowskii replied:
Окей, сделал `pop_front` сверху

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,309 @@
+
+    pub(crate) fn add_message_for_processing(
+        &self,
+        message_with_ack: TopicWriterMessageWithAck,
```
`TopicWriterMessageWithAck` похоже что используется только тут и только чтобы передать пару сообщения и ack_waiter-а. Предлагаю в таких случаях просто передавать два параметра

sultanowskii replied:
Ммм да, до этого еще в одном месте было, в процессе поменял, осталось. Убрал

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,309 @@
+            metadata_items: vec![],
+            data: message.data,
+            uncompressed_size: data_size,
+            partitioning: Some(message_data::Partitioning::MessageGroupId(
```
Сервер сейчас не поддерживаем маршрутизацию на уровне сообщений. Мы один раз задаём producer_id на старте, а отдельно у каждого сообщения этого сделать нельзя. Если оно сейчас работает для producer_id, то случайно.

Тут лучше ставить None.

sultanowskii replied:
Поставил `None`

rekby commented on ydb/src/client_topic/topicwriter/stream_writer.rs:
```diff
@@ -0,0 +1,294 @@
+
+impl StreamWriter {
+    pub async fn new(
+        params: StreamWriterParams,
```
Тут StreamWriterParams похоже используется только 1 раз, зачем тут нужна структура - чем это лучше отдельных параметров?

sultanowskii replied:
Делал по аналогии с другими местами, потом понял, что там оно нужно из-за того, что линтер ругается на большое число аргументов

Поправил

rekby commented on ydb/src/client_topic/topicwriter/stream_writer.rs:
```diff
@@ -0,0 +1,294 @@
+        let message_receive_loop_message_queue = params.message_queue.clone();
+        let message_receive_loop_reception_queue = confirmation_reception_queue.clone();
+
+        let writer_loop = tokio::spawn(async move {
```
функции циклов отправки и получения сообщений лучше выносить прямо отдельными функциями, вроде send_loop, receive_loop.

Большие лямбды сложно читаются и сама функция получается очень длинной - её сложно осознать в голове.

Большая вложенность ещё затрудняет понимание когда одна лямбда начинается, другая заканчивается. IDE тоже не показывают такие логические границы.

sultanowskii replied:
Вынес, справедливо

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,309 @@
+                                trace!("Channel error: {}", chan_err);
+                                let _ = stream_writer.stop().await;  // TODO: handle error
+                                helper.set_writer_state(TopicWriterState::Reconnecting);
+                                break;
```
Тут и рядом: break получается очень далеко от loop, даже не на одном экране.

При больших расстояниях я бы в первую очередь подумал о том как можно код упростить и написать понятнее, в крайнем случае - ставить циклам метки и указывать метку, чтобы было оченвидно какой цикл завершаешь.

Если в начале ещё можно это понять, то потом легко при модификациях добавить ещё какой-то цикл и тогда завершаться будет не то что имелось ввиду изначально.

sultanowskii replied:
Упростил. Повыносил логику, сделал покороче, сейчас выглядит проще

sultanowskii committed:
fix message queue related issues

sultanowskii committed:
tokio mutex

sultanowskii committed:
fix reconnector related issues

sultanowskii committed:
disable partitioning

sultanowskii committed:
remove StreamWriterParams

sultanowskii committed:
extract loops into separate functions

sultanowskii committed:
remove async move block wrapping

sultanowskii committed:
refactor reconnector

sultanowskii committed:
close message queue in stop(), not in flush()

sultanowskii committed:
rename guard to more readable name

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,593 @@
+    fn acknowledge_message(&mut self, seq_no: i64) -> YdbResult<()> {
+        let Some(message) = self.messages.pop_front() else {
+            return Err(YdbError::Custom(format!(
+                "ack unexpected message with seq_no={}",
```
Нужно сообщения об ошибках делать разными, чтобы потом понимать что именно произошло.

Например тут надо написать что очередь сообщений пуста. Иначе непонятно как потом по логам различать ситуацию когда у тебя нет сообщений от того что пришёл неправильный номер (как в соседней ветке).

sultanowskii replied:
Резонно, поправил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,593 @@
+        }
+
+        self.last_sent_index = match self.last_sent_index {
+            Some(idx) if idx > 0 => Some(idx - 1),
```
Тут непонятно почему last_sent_index уменьшается. Что тогда означает это поле если не порядковый номер отправленных сообщений?

sultanowskii replied:
Более не актуально

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,340 @@
+}
+
+pub(crate) struct Reconnector {
+    state: Arc<TokioMutex<ReconnectorState>>,
```
Зачем тут нужно разделять поля по mutex-ам?

Обычно проще хранить одно состояние под общим mutex-ом. Сейчас можно случайно получить дедлоки, код более многословный и т.п.

Плюс может быть логическая гонка при изменении разными потоками разных частей состояния, а когда всё под одним мютексом - логика заметно упрощается.

sultanowskii replied:
Тут я еще подумаю

sultanowskii replied:
Сделал

rekby replied:
Пишешь, что сделал, а в коде осталось 2 мютекса

sultanowskii replied:
ну потому что это два разных состояния, показалось излишним

занесу под один

rekby commented on ydb/src/client_topic/topicwriter/stream_writer.rs:
```diff
@@ -0,0 +1,304 @@
+            .send(stream_write_message::FromClient {
+                client_message: Some(ClientMessage::WriteRequest(WriteRequest {
+                    messages: messages_to_send,
+                    codec: 1,
```
Тут нужно использовать именованную константу, можно из протобафа

sultanowskii replied:
Вставил константу

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,340 @@
+                seq_no,
+                created_at: Some(ydb_grpc::google_proto_workaround::protobuf::Timestamp {
+                    seconds: message.created_at.duration_since(UNIX_EPOCH)?.as_secs() as i64,
+                    nanos: message.created_at.duration_since(UNIX_EPOCH)?.as_nanos() as i32,
```
> duration_since(UNIX_EPOCH)

в наносекундах - вернём время с начало эпохи в наносекундах, а не дробную часть от секунд, в итоге время будет неправильным

sultanowskii replied:
Ух, да, поправил

rekby commented on ydb/src/retry.rs:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    let duration_milliseconds = 2u64
+        .pow(attempt as u32)
```
Тут хорошо бы защититься от переполнения, например: `attempt.min(63), до вызова .pow()`

sultanowskii replied:
Добавил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,593 @@
+
+    pub(crate) async fn reset_progress(&self) {
+        let mut inner = self.inner.lock().await;
+        inner.reset_progress()
```
Где-то в этом месте похоже что нужно сбрасывать ещё и очередь ожидаемых ack-ов.

sultanowskii committed:
improve message queue by replacing last sent index with a second deque

sultanowskii committed:
make flush only wait for messages that have been there at the beginning of operation, update tests

sultanowskii committed:
propogate non-retriable error to TopicWriter, notify all listeners

sultanowskii committed:
match-related code improvements

sultanowskii committed:
fix review issues

sultanowskii committed:
group mutex-protected fields under single mutex

sultanowskii committed:
style

sultanowskii committed:
refactor reconnection loop

sultanowskii committed:
minor style changes

sultanowskii committed:
post-rebase fix

sultanowskii committed:
return fatal error

sultanowskii committed:
fix several bugs

sultanowskii committed:
draft

sultanowskii committed:
improve

sultanowskii committed:
improve test

sultanowskii committed:
minor fixes, remove local file

sultanowskii committed:
style

sultanowskii committed:
remove excessive comment

sultanowskii committed:
style/minor code fixes

sultanowskii committed:
simplify

sultanowskii committed:
fix several message queue bugs, improve implementation

sultanowskii committed:
more message queue tests

sultanowskii committed:
simplify test helper, turn to blackbox testing where possible

sultanowskii committed:
remove useless, add tests

sultanowskii committed:
fix linter issue

sultanowskii committed:
add comment

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+                }
+                Some(seq_no) = message_acknowledged_rx.recv() => {
+                    if seq_no == last_seq_no {
+                        break;
```
А если нам придёт ack с другим номером? По ошибке у нас или на сервере?
Тут сообщение просто игнорируется, а потом мы ака на него уже никогда не получим, т.к. аки не повторяются.

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+    fn check_message_seq_no(&self, seq_no: i64) -> YdbResult<()> {
+        match self.last_added_seq_no {
+            Some(last_added_seq_no) if seq_no <= last_added_seq_no => Err(YdbError::custom(
+                format!("message with seq_no={seq_no} is not newer than the last written message",),
```
переменную часть сообщения лучше ставить в конце - чтобы потом текст удобно было грепать:

```
message seq_no is not greater than the last written message: seq_no={seq_no}
```

sultanowskii replied:
Поправил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+
+    fn would_buffer_exceed_threshold_with_item(
+        buffer: &[MessageData],
+        _item: &MessageData, // Задел на будущее
```
во внутренних API задел оставлять не нужно: сейчас это лишние данные хоть и мелкие. А добавка может и не потребуется никогда.

С другой стороны если потребуется - добавить будет просто.

И я не увидел тут .self - эта функция точно должна быть методом структуры?

sultanowskii replied:
Убрал, упростил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+        let Some(message) = self.messages.front() else {
+            return AppendMessageToSendBufferResult::CouldNotGetMessage;
+        };
+        if MessageQueueInner::would_buffer_exceed_threshold_with_item(
```
А зачем мы тут ждём переполнения вместимости, чтобы сказать что буфер заполнен?

Получается сейчас чтобы буфер отправился нужно theshold + 1 сообщений, причём +1 останется неотправленным.

Что если возвращать ::Full сразу после добавления в буфер последнего элемента, чтобы не ждать ещё +1?

sultanowskii replied:
Была идея задела. Поправил

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+    fn acknowledge_message(&mut self, seq_no: i64) -> YdbResult<()> {
+        let Some(message) = self.sent_messages.pop_front() else {
+            return Err(YdbError::custom(format!(
+                "ack unexpected message with seq_no={seq_no}: queue is empty",
```
переменную часть в конец

sultanowskii replied:
Унес

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+
+        if message.seq_no != seq_no {
+            return Err(YdbError::custom(format!(
+                "ack unexpected message with seq_no={seq_no}: message seq_no mismatch",
```
переменную часть в конец

sultanowskii replied:
Унес

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+            )));
+        }
+
+        self.message_acknowledged_tx.send(seq_no).unwrap();
```
Тут лучше вернуть ошибку, чем делать unwrap.

Хотя врядли message_acknowledged_rx закроется раньше, т.к. сейчас кода закрытия нет вообще. Но в будущем это может измениться, да и вообще unwrap в продакшн скорее плохо, особенно когда можно простым способом вернуть ошибку.

sultanowskii replied:
Сделал

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+
+        let q2 = Arc::clone(&q);
+        let h2 = tokio::spawn(async move {
+            q2.get_messages_to_send(10, Duration::from_millis(500))
```
Поскольку тут всегда ждём до конца таймаута (сообщений нехватает) - лучше таймаут ставить поменьше, чтобы не сильно тормозить тесты. Я обычно ставлю 50-100мс

sultanowskii replied:
Уменьшил

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,376 @@
+
+        let duration = message.created_at.duration_since(UNIX_EPOCH)?;
+        self.message_queue
+            .add_message(MessageData {
```
Конвертацию типов сообщений лучше делать через impl From

sultanowskii replied:
Сделал

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,376 @@
+
+                ReconnectionLoopStatus::WaitForErrorOrCancellation(error_receiver)
+            }
+            Err(err) => {
```
А если ошибка ретраибельная?
Нужно чтобы общий цикл реконнекта ловил в т.ч. ошибки при самом реконнекте.

sultanowskii replied:
Не понял. Машина состояний, разбором ошибок занимается состояние HandleError. Поймали ошибку при переподключении - переход в HandleError.

rekby replied:
Тут такая проблема: при первой попытке подключения сервер оказывается недоступен и в init_tx отправляется ошибка.

loop уровнем сделает ретрай, но код создания писателя в "match init_rx.await" уже получит ошибку.

в init_rx нужно отправлять ответ только когда попытки подключения завершились либо успехом, либо не-ретраибельной ошибкой.

Кстати можно через него-же отправлять и seq_no после первой попытки подключения. Это (вкупе с другими перемещениями) поможет избежать общего состояния в ReconnectionHelper.state.

rekby commented on ydb/src/errors.rs:
```diff
@@ -253,7 +253,7 @@ impl YdbError {
                 use tonic::Code;
                 match status.code() {
-                    Code::Aborted | Code::ResourceExhausted => NeedRetry::True,
+                    Code::Aborted | Code::ResourceExhausted | Code::Unknown => NeedRetry::True,
```
Почему нужно ретраить неизвестные ошибки? Причём всегда

sultanowskii replied:
Потому что во время локальных тестов я ловил вот такое:

```
Error: TransportGRPCStatus(Status { code: Unknown, message: "interceptor transport error: tonic::transport::Error(Transport, hyper::Error(Io, Kind(ConnectionReset)))", source: Some(interceptor transport error: tonic::transport::Error(Transport, hyper::Error(Io, Kind(ConnectionReset)))) })
```

Видно, что это Unknown ошибка. По сообщению грепать?

rekby replied:
интересная ошибка, не видел такой. 
Неизвестные ошибки если хочется ретраить лучше возвращать как "ретраибельные для идемпотентных операций".

sultanowskii replied:
Сделал

rekby commented on ydb/src/topics_test.rs:
```diff
@@ -1274,3 +1274,144 @@ async fn read_batch_merges_and_respects_hard_limit() -> YdbResult<()> {
+            .await?;
+    }
+
+    proxy.set_allow_forward(false);
```
Это только делает паузу в обработке новых подключений, но не рвёт существующие соединения, так что теста реконнекта тут не происхоидт.

sultanowskii replied:
Стоп, разве?

Документация `copy_bidirectional`:

```
The future will complete successfully once both directions of communication has been shut down.
```

sultanowskii replied:
\+ При запуске в логах можно увидеть, что соединение рвется

```
Error, trying to reconnect: TransportGRPCStatus(Status { code: Unknown, message: "interceptor transport error: tonic::transport::Error(Transport, hyper::Error(Io, Kind(ConnectionReset)))", source: Some(interceptor transport error: tonic::transport::Error(Transport, hyper::Error(Io, Kind(ConnectionReset)))) })
```

Ради интереса еще запустил этот тест на ветке `master`, тест упал, словил:

```
2026-05-12T18:26:39.371622Z  WARN topic_writer_reconnects: ydb::client_topic::topicwriter::writer: error receive message for topic writer receiver stream loop: TransportGRPCStatus(Status { code: Unknown, message: "h2 protocol error: error reading a body from connection", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) })
Error: TransportGRPCStatus(Status { code: Unknown, message: "h2 protocol error: error reading a body from connection", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) })
test topics_test::topic_writer_reconnects ... FAILED
```

Видно, что речь про это место: https://github.com/sultanowskii/ydb-rs-sdk/blob/master/ydb/src/client_topic/topicwriter/writer.rs#L185, т.е. не при создании подключения, а уже в работе

rekby commented on ydb/src/client_topic/topicwriter/writer_reception_queue.rs:
Нужно убрать использование unwrap везде в коде писателя. Сейчас unwrap может упасть при закрытии писателя (может и не упасть, но это надо доказывать и потом всегда это свойство сохранять).

sultanowskii replied:
Убрал, unwrap остались только в тестах

rekby commented on ydb/src/client_topic/topicwriter/stream_writer.rs:
```diff
@@ -0,0 +1,304 @@
+            writer_state.connection_info = ConnectionInfo {
+                partition_id: init_response.partition_id,
+                session_id: init_response.session_id,
+                last_seq_no_assigned: init_response.last_seq_no,
```
last_seq_no_assigned из init-сообщения нужно брать только при первом подключении.

На реконнектах оно должно сохранять своё значение или будет проблема если в момент реконнекта во внутреннем буфере будут неподтверждённые сообщения, то автоматические номера начнут назначаться повторно.

sultanowskii replied:
Сделал

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,376 @@
+            let mut writer_state = self.writer_state.lock().await;
+            writer_state
+                .confirmation_reception_queue
+                .add_ticket(TopicWriterReceptionTicket::new(seq_no, reception_type));
```
ВАЖНО:

Тут получается гонка между добавлением тикета и ack-ом сообщения от сервера.

Если сервер сработает быстро, а этот код притормозит - можно получить ack о сообщении ещё до того как тикет добавится в confirmation_reception_queue и свалиться с 

```
                        let Some(expected_seq_no) = expected_seq_no else {
                            return Err(YdbError::custom(
                                "Expected reception ticket to be actually present",
                            ));
                        };
```

sultanowskii replied:
Теперь под одним мьютексом, "обгона" происходить не должно

rekby commented on ydb/src/client_topic/topicwriter/message_queue.rs:
```diff
@@ -0,0 +1,585 @@
+    inner: Arc<TokioMutex<MessageQueueInner>>,
+
+    new_message_added: Arc<Notify>,
+    message_acknowledged_rx: Arc<TokioMutex<mpsc::UnboundedReceiver<i64>>>,
```
Тут будет утечка памяти: в канал складываются все подтверждения, а вычитываются только на flush(). Пользователь НЕ обязан вызывать flush().

Нужно заменить очередь подтверждений на хранение seq_no последнего подтверждённого сообщения + отдельно уведомления о том, что значение обновилось, чтобы можно было подождать нужного значения.

sultanowskii replied:
Поменял на Notify

rekby commented on ydb/src/client_topic/topicwriter/writer.rs:
```diff
@@ -1,448 +1,293 @@
         };
 
+        self.message_queue
+            .wait_for_messages_to_be_acknowledged(&self.cancellation_token)
```
А нужно это ожидание? 

может хватить только следующего:

```
        Ok(flush_op_completed.await?)
```

?

sultanowskii replied:
По ощущениям все же хотим дождаться этой очереди.

Если считаешь иначе - прошу пояснить, почему `Ok(flush_op_completed.await?)` достаточно

rekby commented on ydb/src/client_topic/topicwriter/stream_writer.rs:
```diff
@@ -0,0 +1,304 @@
+        let cancellation_token = CancellationToken::new();
+
+        // Both loops share the same oneshot error channel.
+        let shared_error_tx = Arc::new(TokioMutex::new(Some(error_tx)));
```
будет ли тут удобнее `tokio::sync::SetOnce`?

sultanowskii replied:
Посмотрел - не увидел прямого выигрыша + для этого нужно обновить версию либы и пакета, как я понял

sultanowskii committed:
fix review issues

sultanowskii committed:
rename

sultanowskii committed:
simplify message queue

sultanowskii committed:
fix linter issues

sultanowskii committed:
wip refactor

sultanowskii committed:
tests

sultanowskii committed:
remove Inner suffix from MessageQueue

sultanowskii committed:
remove println

sultanowskii committed:
set last_seq_no_assigned only once

sultanowskii committed:
fmt

rekby commented on ydb/src/client_topic/topicwriter/queue.rs:
```diff
@@ -0,0 +1,534 @@
+        };
+        if write_ack.seq_no != expected_seq_no {
+            return Err(YdbError::custom(format!(
+                "reception ticket and write ack seq_no mismatch: expected_seq_no: {}, actual_seq_no: {}",
```
```suggestion
                "reception ticket and write ack seq_no mismatch: expected_seq_no: {expected_seq_no}, actual_seq_no: {write_ack.seq_no}"
```

тут перепутаны местами подписи и значения. Если записывать значения прямо внутрь строк - такие опечатки будут более заметны

sultanowskii replied:
```
invalid format string: field access isn't supported
not supported in format stringrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20[0]?0#file:///Users/sultanowskii/projects/sultanowskii/ydb-rs-sdk/ydb/src/client_topic/topicwriter/queue.rs)
queue.rs(211, 118): consider using a positional formatting argument instead: `0`, `, write_ack.seq_no`
```

Вынесу во временную переменную

rekby commented on ydb/src/client_topic/topicwriter/writer_options.rs:
Вот тут поля:

```
    pub(crate) path: String,
    pub(crate) producer_id: Option<String>,
    pub(crate) write_request_messages_chunk_size: usize,
    pub(crate) write_request_send_messages_period: Duration,
    pub(crate) auto_set_seq_no: bool,
    pub(crate) flush_timeout: Duration,

    cancellation_token: CancellationToken,
```

стали лишними - после твоих доработок они уже не используются и их надо удалить, как и `#[allow(dead_code)]` - чтобы компилятор ругался если найдёт что-то лишнее.

Часть полей были мёртвыми уже в мастере, так что это скорее косметика "вокруг": при доработках в реконнектор уехали auto_set_seq_no, cancellation_token и вот их точно стоило удалить из основной структуры.

sultanowskii replied:
О, да, это хорошее замечание, правда ты текстом написал про `Writer`, а фрагмент кода указывает на `TopicWriterOptions`.

Первое (`Writer`) поправлю, второе не трогаем ведь, оставляя dead_code и ряд неиспользуемых (пока что) полей в духе `custom_encoders`, `codec`?

rekby commented on ydb/src/client_topic/topicwriter/queue.rs:
```diff
@@ -0,0 +1,534 @@
+        message: MessageData,
+        ack: Option<oneshot::Sender<YdbResult<MessageWriteStatus>>>,
+    ) -> YdbResult<()> {
+        let reception_type = ack.map_or(
```
Тут получается что сначала ack создается как Option-тип, а в середине меняется на enum. Это всё происходит во внутреннем API, так что сохранение старого кода смысла не несёт.

Давай тут перейдём к одному единому варианту:
либо вернём все на Option (текущий enum ему эквивалентен), либо тогда уже всё на enum с самого начала - в TopicWriter.write_message.


Мне больше нравится Option, т.к. это стандартный тип и у нас нет большого количества его использований, чтобы повышать читаемость None-варианта (оно ровно одно). Но тут я не настаиваю - можно сделать любой из вариантов. Важно чтобы было единообращие по всему коду, без конвертаций.

sultanowskii replied:
Резонно, переделал на option

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,443 @@
+        let queue = Queue::new();
+        let state = Arc::new(TokioMutex::new(ReconnectorState {
+            status: ReconnectorStatus::Working,
+            loop_handle: None,
```
loop_handle не используется внутри Helper-а.

Если переместить его в Reconnector, то можно избавиться от Option

sultanowskii replied:
Сделал

rekby commented on ydb/src/client_topic/topicwriter/reconnector.rs:
```diff
@@ -0,0 +1,443 @@
+    }
+}
+
+struct ReconnectorState {
```
тут можно:
- loop_handle: не нужен внутри ReconnectorLoop, можно оставить отдельным полем в Reconnector
- queue - сам по-себе легко-клонируемый и синхронизирован внутренними механизмами - его можно передать просто как параметр и получать доступ к очереди без доп. мютекса.
- connection_info: по него из стрима наружу уходит last_seq_no_assigned. Если его передавать через init_tx, то поле тоже получается ненужным. К тому же оно должно передаваться как раз ровно 1 раз: на начальном подключении. При переподключениях его можно уже не передавать и не спрашивать. Нумерация не должна сбрасываться при переподключениях, а должна продолжаться последовательно с самого начала. Если на переподключении seq_no будет меньше, то это повод переотправить недостающие сообщения (можно просто переотправлять все, которые в буфере сохранились - это тоже будет работать), а не изменить нумерацию входящих сообщений, на которые влияет connection_info.last_seq_no_assigned. Т.е. seq_no нужно передать наверх один раз - при первой успешной попытке подключения. Потом счётчик нумерации становится автономным.
- status: можно заменить на tokio::sync::watch, с пишущей частью в reconnectorLoop и читающей в реконнекторе. Там можно будет и подписаться на его изменение, чтобы узнать о том что реконнектор закрылся на всегда и текущее состояние читать.


Сейчас проблема в том, что одно и то же состояние шарится между двумя разными структурами, каждая со своим назначением. Этого следут избегать покуда есть возможность, т.к. шаренное состояние сложно понимать и отлаживать.

sultanowskii replied:
Справедливые замечания, поправил по ним, код подупростился, спасибо

rekby commented on ydb/src/client_topic/topicwriter/queue.rs:
```diff
@@ -0,0 +1,534 @@
+            return;
+        };
+
+        loop {
```
Тут похоже что ожидание зависнет если были хоть какие-то сообщения (в last_seq_no уже есть сообщения), но сейчас в очереди на отправку ничего нет.

Предположительный сценарий:
```
writer.write_message()
writer.flush()
writer.flush()
```

sultanowskii replied:
После переезда на Queue некорректно вынес логику последнего добавленного seq_no. Сейчас переложил ответственность на MessageQueue, дописал тест.

Поправил

sultanowskii committed:
refactor state, fix other issues

sultanowskii committed:
clarify error, update test

sultanowskii committed:
fix wait_for_messages_to_be_acknowledged hanging after queue is fully drained

sultanowskii committed:
improve style, remove unnecessary modifiers, extract functions, name things more consistently

rekby committed:
Merge branch 'master' into topic-writer-retry

rekby committed:
Address review feedback on TopicWriter retry

- writer_options.rs: drop unused fields and types
  (TopicWriterConnectionOptions, TopicWriterRetrySettings,
  auto_created_at, codec, custom_encoders, connection_options),
  remove #[allow(dead_code)] so future dead fields surface
- reception_queue.rs: support concurrent flush() calls by replacing
  the single flush_finished_sender + flush_flag on tickets with
  a list of pending flushes keyed by threshold seq_no
- message_queue.rs: remove unwrap() in append_message_to_send_buffer
- retry.rs: move misplaced try_vec_* tests to types_converters.rs
